### PR TITLE
Feature: improve offsite donation redirect flow in new gateway api

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves https://github.com/impress-org/givewp/issues/6186

## Description

This pull request adds updates to the new payment gateway logic in favor of the offsite payment gateway. Offsite type payment gateway only responsible for the offsite payment return callback handler and the core/framework will responsible for processing commands and redirects.

Important updates:
- `LegacyPaymentGatewayAdapter` has been updated to pass `OffsiteGatewayPaymentData` to offsite payment gateway type. The only difference, it the container redirects, failed, and cancelled the gateway route URL. Three callback has been added to the `OffsitePaymentGateway` contract to define default callback for these gateway routes.

- We need two types of gateway route URLs:
   - Payment gateway must visit them within a limited time like offsite payment redirect
   - Payment gateway will visit any update to payment on the gateway server.
     I updated the `GenerateGatewayRouteUrl` class to generate URL with nonce and this once will be validated in `GatewayRoute`.
- An action `ProcessOffsitePaymentRedirectOnGatewayRoute` has been added to handle offsite payment return in `GatewayRoute`.

## Testing
Test offsite payment flow:
1. Setup a second website (http://freshdb.test).
2. Add below code as in mu plugin.
```php
        add_action( 'init', function(){
             if( ! isset( $_GET['test-offsite-redirect'] ) ) {
                return;
             }
        
             $redirect = urldecode($_GET['redirect']);
        
             wp_redirect($redirect);
            exit();
        });
```
3. Process payment with "Test Gateway Offsite"
    If you want to test failed or cancelled donation flow then change url to:
```$offsiteGatewayPaymentData->failedRedirectUrl or $offsiteGatewayPaymentData->cancelledRedirectUrl```

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit and/or end-to-end tests
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

File: src/Framework/PaymentGateways/Routes/GatewayRoute.php [hunk 1]
```diff
@@ -2,77 +2,109 @@
...
+            throw new PaymentGatewayException('This route does not have valid nonce.');
+        }
 
-            /** @var PaymentGateway $gateway */
-            $gateway = give($paymentGateways[$data->gatewayId]);
+        /** @var PaymentGateway $gateway */
+        $gateway = give($paymentGateways[$data->gatewayId]);
 
-            $allowedGatewayMethods = $gateway->routeMethods;
+        $allowedGatewayMethods = Gateway::isOffsitePaymentGateway($gateway) ? $gateway->routeMethods : [];
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/Routes/GatewayRoute.php:57]: @ravinderk gateways do not need to implement the offsite interface to have routeMethods.  please revert this change.

File: src/Framework/PaymentGateways/Routes/GatewayRoute.php [hunk 2]
```diff
@@ -2,77 +2,109 @@
...
+         * For this reason we need core to involve to handle redirect.
+         */
+        if (in_array($data->gatewayMethod, OffsiteGatewayInterface::defaultRouteMethods, true)) {
+            (new ProcessOffsitePaymentRedirectOnGatewayRoute($gateway))
+                ->handleGatewayRouteMethod($data->donationId, $data->gatewayMethod);
+        } else {
             $gateway->handleGatewayRouteMethod($data->donationId, $data->gatewayMethod);
         }
+
+        exit();
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/Routes/GatewayRoute.php:91]: @ravinderk I think we might be overcomplicating this whole thing.  We moved away from trying to compensate for every payment status command for this route handler in the latest version of the gateway API.

File: src/Framework/PaymentGateways/Types/OffSitePaymentGateway.php [hunk 3]
```diff
@@ -2,14 +2,53 @@
...
+     * Handle cancelled donation redirect.
+     *
+     * @param int $donationId
+     *
+     * @return PaymentCancelled
+     */
+    public function returnCancelFromOffsiteRedirect($donationId)
+    {
+        return new PaymentCancelled();
+    }
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/Types/OffSitePaymentGateway.php:53]: @ravinderk there was a reason we had an `OffsiteGateway` trait. It's because a gateway should have the flexibility to extend OffSitePaymentGateway, or implement OffsiteGatewayInterface themselves and pull in the trait for the helpers.

Also same comment about moving away from all payment status commands.  We need to talk through this.

File: src/PaymentGateways/Gateways/TestGateway/Commands/CreateTestGatewayOffsitePaymentUrlCommand.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        // 3) Process payment with "Test Gateway Offsite"
+        //
+        // If you want to test failed or cancelled donation flow then change url to:
+        // $offsiteGatewayPaymentData->failedRedirectUrl or $offsiteGatewayPaymentData->cancelledRedirectUrl
+        return sprintf(
+            '%1$s?test-offsite-redirect=1&redirect=%2$s',
+            'http://freshdb.test',
+            urlencode($offsiteGatewayPaymentData->failedRedirectUrl)
+        );
+    }
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/TestGateway/Commands/CreateTestGatewayOffsitePaymentUrlCommand.php:48]: - This would not be considered a command, probably an action if anything.  
- I don't think we need those comments.
- use a wp function for getting home_url() instead of hardcoding domains
- Limit this scope of actions, you are only using one property here so just make that the param instead of the entire DTO

File: src/PaymentGateways/DataTransferObjects/OffsiteGatewayPaymentData.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            GenerateGatewayRouteUrl::class,
+            $self->gatewayId,
+            'returnCancelFromOffsiteRedirect',
+            $self->donationId,
+            null,
+            [ 'withNonce' => true ]
+        );
+
+        return $self;
+    }
```
Comment (jonwaldstein) [inline src/PaymentGateways/DataTransferObjects/OffsiteGatewayPaymentData.php:72]: I don't think this should be it's own DTO. I would just create the urls when needed.

File: src/Framework/PaymentGateways/Routes/GatewayRoute.php [hunk 6]
```diff
@@ -86,4 +118,29 @@ private function isValidListener()
...
+            return wp_verify_nonce(
+                $data->nonce,
+                (new GenerateGatewayRouteUrl())->getNonceActionName($data)
+            );
+        }
+
+        // We can return true for other gateway route urls like webhook, paypal ipn
+        // Because these routes will not contain nonce.
+        return true;
+    }
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/Routes/GatewayRoute.php:145]: Let's think through the best way for including and handling nonces in the route.   It's a little weird that this is completely optional and only validates when it's included.

File: src/Framework/PaymentGateways/Actions/GenerateGatewayRouteUrl.php [hunk 7]
```diff
@@ -2,18 +2,24 @@
...
+     * @param string $gatewayId
+     * @param string $gatewayMethod
+     * @param int $donationId
+     * @param array|null $additionalQueryArgs
+     * @param array|null $optParams
+     *
+     * @return string Url.
      */
-    public function __invoke($gatewayId, $gatewayMethod, $donationId, $args = null)
+    public function __invoke($gatewayId, $gatewayMethod, $donationId, $additionalQueryArgs = null, $optParams = null)
```
Comment (jonwaldstein) [inline src/Framework/PaymentGateways/Actions/GenerateGatewayRouteUrl.php:22]: As previously commented, we need to think through properly including & validating nonces.  This would not be clear enough if it's something we need to do.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]
@ravinderk appreciate the effort on this. I see where you're going with these changes but we need to think through a lot of this together before implementing. It seems like the complexity is starting to pile up on the gateway api that might not actually be necessary.  Let's get together and figure out the problems your trying to solve before modifying this PR.

**Comment by ravinderk** [comment]
Closing in favor of https://github.com/impress-org/givewp/pull/6196
